### PR TITLE
Add PyTorch operator tests

### PR DESCRIPTION
**Pull request by amd-eochoalo**
This PR adds pytorch operator tests. This follows a similar approach to ONNX ops, where the tests are defined at a higher level of abstraction (in this case the tests are defined at the PyTorch level) and pre-generated ahead of time. The workflow is as follows:

1. Add a new test definition in `generate_test.py`
2. Regenerate the tests `python generate_test.py`
3. Run the tests with `pytest`.

The MLIR representation is generated AOT of time and saved to a file. Numpy files are used to store the input parameters and the expected output. A json file is used to bundle these parameters together along with rtol and atol which are used to compare the expected output with the observed output.

### CODE REVIEW

**Comment by amd-eochoalo** [comment]
While writing these tests it looks like one bug has been discovered.

```python
    def forward(self, *args):
        return torch.ops.aten.triu_indices(8, 8, 0)
```

 result[0]: metadata is 2x36xi64; expected that the view matches 2x36xsi64. Notice signless vs signed integers.

File: pytorch_ops/example_test_generation.py [hunk 1]
```diff
@@ -0,0 +1,503 @@
+import iree.runtime as ireert
+import numpy as np
+import iree.turbine.aot as aot
+import torch
+from pathlib import Path
+
+
```
[the remark this answers is not in the document]
Comment (amd-eochoalo) [reply pytorch_ops/example_test_generation.py:7]: Include multiple tests in a single file. Fx export (look at other PR)

File: pytorch_ops/conftest.py [hunk 2]
```diff
@@ -0,0 +1,461 @@
+# Copyright 2024 The IREE Authors
```
Comment (amd-eochoalo) [inline pytorch_ops/conftest.py:1]: TODO: Remove these files, these are identical to ONNX ops conftest.py

File: pytorch_ops/example_test_generation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (amd-eochoalo) [reply pytorch_ops/example_test_generation.py]: @Groverkss are you sure about wanting to include multiple tests in a single file? I am a bit against it since it complicates a couple of things. Nothing too major, but I think it would make testing less obvious. E.g.,

* If compilation fails for one of these multiple tests in a single file, we wouldn't immediately know which one.
* I originally experimented with using a single json file for storing which function corresponds to which inputs and expected results. But this way one can't reuse the `--flagfile` for iree-run-module. Maybe multiple flagfiles per module would work better.
* The ONNX setup to run tests is to compile and run as a single test. I think this can be modified such that compilation is a pre-step and then we run multiple times to separate different reporting from pytest. But at this point, why would we want this? 

Also, you mentioned that I shouldn't write tests for all ops and just focus on GEMM and convolutions. Would you consider that the end point of this story? I have tests for ~40 ops in my original branch.

[the remark this answers is not in the document]
Comment (Groverkss) [reply pytorch_ops/example_test_generation.py]: fair enough, we dont have to support multiple tests in a single file.

File: pytorch_ops/generate_tests.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        self.save_inputs(*inputs)
+        self.save_results(*expected_results)
+        self.save_config()
+
+    def generate_expected_value(self, *args):
+        results = self.forward(*args)
+        return [results]
+
+    def run_turbine(self, *args):
+        opt_module = torch.compile(self, backend="turbine_cpu")
```
[the remark this answers is not in the document]
Comment (amd-eochoalo) [reply pytorch_ops/generate_tests.py:93]: I'm using run_turbine here to make sure that the expected_output from pytorch matches the observed_output when generating the test. We can possibly remove this step. This step happens during the generation of the tests, but it will also happen during the test itself. E.g.,

```python
def generate_tests():
  expected = run_torch()
  observed = run_turbine
  assert expected == observed # we make sure that this is true before even writing the test

def run_test():
  expected = load_expected()
  observed = compile_and_run_with_ireert()
  assert expected == observed
```

This is a little bit useful as it avoids spending time generating tests which will fail either way locally.

[the remark this answers is not in the document]
Comment (amd-eochoalo) [reply pytorch_ops/generate_tests.py:93]: Also, I am using aot.export above. Do you want me to use torch.export instead?

[the remark this answers is not in the document]
Comment (Groverkss) [reply pytorch_ops/generate_tests.py:93]: Oh sorry for the miscommunication. aot.export/torch.export are same to me haha. I meant aot.export, you are correct.

File: pytorch_ops/generate_tests.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+def camel_to_snake(name):
+    result = ""
+    for i, char in enumerate(name):
+        if char.isupper() and i > 0:
+            result += "_"
+        result += char.lower()
+    return result
+
+
+class TestGenerator(torch.nn.Module, ABC):
```
Comment (Groverkss) [inline pytorch_ops/generate_tests.py]: This is mostly well written, well done!

File: pytorch_ops/generate_tests.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    for cls in [AB]:
+        random = RandomSession()
+        inputs = (random.rand(64, 64, dtype=dtype), random.rand(64, 64, dtype=dtype))
+        dyn_dim = torch.export.Dim("N")
+        dynamic_shapes = {"left": {0: dyn_dim}, "right": {1: dyn_dim}}
+        instance = cls(
+            *inputs,
+            name=cls.__name__ + str(dtype),
+            export_kwargs={"dynamic_shapes": dynamic_shapes},
+        )
```
Comment (Groverkss) [inline pytorch_ops/generate_tests.py]: Uhhh, this is not how it should be. Dynamic dimensions should not be added at class initialization time. It should be a class property. Instead implement a `def get_dynamic_shapes` in TestGenerator class and ask the user to override it.

The default implementation can be empty, which is everything is static.

File: pytorch_ops/generate_tests.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        dyn_dim = torch.export.Dim("N")
+        dynamic_shapes = {"left": {0: dyn_dim}, "right": {1: dyn_dim}}
+        instance = cls(
+            *inputs,
+            name=cls.__name__,
+            export_kwargs={"dynamic_shapes": dynamic_shapes},
+        )
+        instance.generate_test(atol=1e-2, rtol=1e-2)
+for dtype in [torch.float32, torch.float16]:
+    for cls in [AB]:
```
Comment (Groverkss) [inline pytorch_ops/generate_tests.py]: These loops are not very scalable. I would prefer if we made each class self contained. A class definition should generate a test on it's own.

You can ask each class to have a seed to use to generate something. I'd say do it pytest style where every  marked function is a test. Something like:

```python
class Matmul(TestGenerator):
  
  @test(seed=42)
  def one(self, seed):
    lhs = torch.rand(64, 64, dtype=dtype)
    rhs = torch.randn(64, 64, dty

  def forward(self, lhs, rhs):
    return lhs @ rhs

  def dynamic_shapes(self):
    dyn_dim = torch.export.Dim("N")
    return {
      "left": {0: dyn_dim},
      "right": {1: dyn_dim},
    }
```

File: pytorch_ops/conftest.py [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+            outs = process.stdout.decode("utf-8")
+        except:
+            outs = str(process.stdout)
+
+        test_github_url = (
+            "https://github.com/iree-org/iree-test-suites/blob/main/onnx_ops/"
+            + cwd.relative_to(THIS_DIR).as_posix()
+        )
+
+        # TODO(scotttodd): handle large sources somehow (truncate, skip, link to source, etc.)
```
Comment (MaheshRavishankar) [inline pytorch_ops/conftest.py]: You are adding a TODO or Scott?

**Comment by MaheshRavishankar** [comment]
Oh I did have one comment. You need to add a README.md (just like the torch-models tests do) to describe this suite.

File: pytorch_ops/generate_tests.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+    seed: int = 0
+
+    def __post_init__(self):
+        self.generator.manual_seed(self.seed)
+
+    def rand(self, *args, **kwargs):
+        if kwargs.get("generator"):
+            raise ValueError("RandomSession uses its own generator")
+        kwargs["generator"] = self.generator
+        return torch.rand(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (amd-eochoalo) [reply pytorch_ops/generate_tests.py]: @bjacob that's correct! I changed it a little bit but the idea is that this is just a convenience wrapper that if it is used then the generator which is stored is used instead of the one passed. I've documented this wrapper encouraging its use.

File: pytorch_ops/conftest.py [hunk 9]
```diff
@@ -0,0 +1,482 @@
...
+            errs = process.stderr.decode("utf-8")
+        except:
+            errs = str(process.stderr)
+        try:
+            outs = process.stdout.decode("utf-8")
+        except:
+            outs = str(process.stdout)
+
+        test_github_url = (
+            "https://github.com/iree-org/iree-test-suites/blob/main/onnx_ops/"
```
Comment (Groverkss) [inline pytorch_ops/conftest.py]: update please

**Review by Groverkss — APPROVED** [review]
LGTM, Very well documented!
